vcd/resource_vcd_vapp_vm.go

```diff
@@ -611,19 +658,62 @@ func resourceVcdVAppVmUpdateExecute(d *schema.ResourceData, meta interface{}) er
 			}
 		}
 
-		if d.Get("power_on").(bool) {
+		if d.HasChange("network") {
+			networkConnectionSection, err := networksToConfig(d.Get("network").([]interface{}), vdc, vapp, vcdClient)
+			if err != nil {
+				return fmt.Errorf("unable to setup network configuration for update: %s", err)
+			}
+			err = vm.UpdateNetworkConnectionSection(&networkConnectionSection)
+			if err != nil {
+				return fmt.Errorf("unable to update network configuration: %s", err)
+			}
+		}
+
+	}
+
+	// If the VM was powered off during update but it has to be powered off
+	if d.Get("power_on").(bool) {
+		vmStatus, err := vm.GetStatus()
+		if err != nil {
+			return fmt.Errorf("error getting VM status before ensuring it is powered on: %s", err)
+		}
+		log.Printf("[DEBUG] Powering on VM %s after update. Previous state %s", vm.VM.Name, vmStatus)
 
+		// Simply power on if customization is not requested
+		if !customizationNeeded && vmStatus != "POWERED_ON" {
 			task, err := vm.PowerOn()
 			if err != nil {
-				return fmt.Errorf("error Powering Up: %#v", err)
+				return fmt.Errorf("error powering on: %s", err)
 			}
-
 			err = task.WaitTaskCompletion()
 			if err != nil {
 				return fmt.Errorf(errorCompletingTask, err)
 			}
 		}
 
+		// When customization is requested VM must be un-deployed before starting it
+		if customizationNeeded {
+			log.Printf("[TRACE] forced customization for VM %s was requested. Current state %s",
+				vm.VM.Name, vmStatus)
+
+			if vmStatus != "POWERED_OFF" {
+				log.Printf("[TRACE] VM %s is in state %s. Un-deploying", vm.VM.Name, vmStatus)
+				task, err := vm.Undeploy()
```

Comment:
Above you already have code which does Undeploy if customization is needed. Why do you do it here again? Or from the other perspective, why do you Undeploy above, if you have this logic here?

Reply:
It is because this resource needs a major refactor at some point in time. Probably even with my next PR for "guest properties".

The main problem here is that `create` method _always_ calls `update` to fully customize the VM (this should not happen as `create` and `update` should be separate operations). The above "undeploy" is only called if some other resource settings have changed during update (like `memory`, `cpu`, `cpu_cores`, `power_on`, `disk`, `expose_hardware_virtualization`, `network`).

But if the user only forces customization `customization.force = true` without altering any other options (listed in my above comment) - then the above undeploy will not be triggered and I need to do it here.